### PR TITLE
Update St1inch.sol

### PR DESCRIPTION
**Pull request by zodenode**
Added Natspec and some inline comments for further clarity

### CODE REVIEW

File: contracts/St1inch.sol [hunk 1]
```diff
@@ -322,6 +328,15 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
         (loss, ret) = _earlyWithdrawLoss(amount, balanceOf(account));
         canWithdraw = loss <= amount * maxLossRatio / _ONE_E9;
     }
+
```
Comment (galekseev) [inline contracts/St1inch.sol:331]: We usually avoid documenting private functions unless they add some additional logic description and improve understanding of inner working. The docs that are generated do not include private functions natspecs. I suggest to either add clear description how and why loss and return amount calculated (better) or drop the natspec for this function (acceptable).

File: contracts/St1inch.sol [hunk 2]
```diff
@@ -322,6 +328,15 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
     }
+    
+    /**
+    * @notice Calculates the loss amount if the staker does early withdrawal at the current block
+    * @dev Internal function
+    * @param depAmount The amount deposited by the staker
+    * @param stBalance The current balance of the staker
+    * @return loss The loss amount
+    * @return ret The return amount
+    */
```
Comment (galekseev) [inline contracts/St1inch.sol]: Unwanted extra line feed

File: contracts/St1inch.sol [hunk 3]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
         }
     }
 
+     /**
```
Comment (galekseev) [inline contracts/St1inch.sol]: Remove natspec for private function

File: contracts/St1inch.sol [hunk 4]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     * @dev Private function to withdraw deposited tokens and burn corresponding staking tokens
+     * @param depositor Struct containing depositor's information
+     * @param balance Amount of staking tokens to be burned
+     */
     function _withdraw(Depositor memory depositor, uint256 balance) private {
-        totalDeposits -= depositor.amount;
-        depositor.amount = 0;
-        // keep unlockTime in storage for next tx optimization
+        totalDeposits -= depositor.amount; // Reduce total deposits by the amount being withdrawn
```
Comment (galekseev) [inline contracts/St1inch.sol]: Redundant obvious comment

File: contracts/St1inch.sol [hunk 5]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+        depositor.amount = 0; // Set depositor's amount to zero
```
Comment (galekseev) [inline contracts/St1inch.sol]: Redundant obvious comment

File: contracts/St1inch.sol [hunk 6]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+        // Update unlockTime in storage for next transaction optimization
```
Comment (galekseev) [inline contracts/St1inch.sol]: Wouldn't verb "store" be better?

File: contracts/St1inch.sol [hunk 7]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
         depositor.unlockTime = uint40(block.timestamp);
-        depositors[msg.sender] = depositor; // SSTORE
-        _burn(msg.sender, balance);
+        depositors[msg.sender] = depositor; // Update depositor's information in storage
```
Comment (galekseev) [inline contracts/St1inch.sol]: SSTORE is a code that means that SSTORE opcode is executed on EVM level. It is important due gas costs. Please do not change this line

File: contracts/St1inch.sol [hunk 8]
```diff
@@ -349,13 +364,18 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+        _burn(msg.sender, balance); // Burn staking tokens
```
Comment (galekseev) [inline contracts/St1inch.sol]: Redundant obvious comment

File: contracts/St1inch.sol [hunk 9]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
 
     // ERC20 methods disablers
 
+     /**
+     * @notice Throws an error indicating that approve functionality is disabled
+     * @dev This function always throws
```
Comment (galekseev) [inline contracts/St1inch.sol]: It is better to express the idea: "Always reverts to disable token transfer ERC-20 feature .

Comment (galekseev) [inline contracts/St1inch.sol]: It is better to use conventional terms: "throws" -> "reverts"

File: contracts/St1inch.sol [hunk 10]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     */
     function approve(address, uint256) public pure override(IERC20, ERC20) returns (bool) {
         revert ApproveDisabled();
     }
 
+    /**
```
Comment (galekseev) [inline contracts/St1inch.sol:407]: The same comments as for approve function

File: contracts/St1inch.sol [hunk 11]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     * @notice Throws an error indicating that transfer functionality is disabled
+     * @dev This function always throws
+     */
     function transfer(address, uint256) public pure override(IERC20, ERC20) returns (bool) {
         revert TransferDisabled();
     }
 
+    /**
```
Comment (galekseev) [inline contracts/St1inch.sol:415]: The same comments as for approve function

File: contracts/St1inch.sol [hunk 12]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     * @notice Throws an error indicating that transferFrom functionality is disabled
+     * @dev This function always throws
+     */
     function transferFrom(address, address, uint256) public pure override(IERC20, ERC20) returns (bool) {
         revert TransferDisabled();
     }
 
+    /**
```
Comment (galekseev) [inline contracts/St1inch.sol:423]: The same comments as for approve function

File: contracts/St1inch.sol [hunk 13]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     * @notice Throws an error indicating that increaseAllowance functionality is disabled
+     * @dev This function always throws
+     */
     function increaseAllowance(address, uint256) public pure override returns (bool) {
         revert ApproveDisabled();
     }
 
+    /**
```
Comment (galekseev) [inline contracts/St1inch.sol:431]: The same comments as for approve function

File: contracts/St1inch.sol [hunk 14]
```diff
@@ -376,23 +396,44 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
+     * @notice Throws an error indicating that decreaseAllowance functionality is disabled
+     * @dev This function always throws
+     */
     function decreaseAllowance(address, uint256) public pure override returns (bool) {
         revert ApproveDisabled();
     }
+
```
Comment (galekseev) [inline contracts/St1inch.sol]: Unwanted extra line feed

**Review by galekseev — CHANGES_REQUESTED** [review]
Please, correct in-code comments

File: contracts/St1inch.sol (hunk 1 above)
Comment (zodenode) [reply contracts/St1inch.sol:331]: Got it, I think the loss and return amount are documented fairly effectively in the `earlyWithdraw` and the `earlyWithdrawTo` Natspec. But I can mimic those too, although it wouldn't provide "additional logic descriptions" so I think the best is to drop the Natspec

**Review by galekseev — APPROVED** [review]

File: contracts/St1inch.sol [hunk 15]
```diff
@@ -322,7 +328,21 @@ contract St1inch is ERC20Pods, Ownable, VotingPowerCalculator, IVotable {
...
-
+    
+    /**
+     * @notice Calculates the loss amount if the staker does early withdrawal at the current block
+     * @dev Internal, Maths:
+     * 1. amount = (deposit - vp(balance)) * K, where
+     * amount - amount to unstake,
+     * deposit - stake amount,
+     * vp(balance) - voting power of the given balance at the current block,
+     * K - koefficient, to adjust the loss amount to 0 at the end of the lock period.
```
Comment (ZumZoom) [inline contracts/St1inch.sol]:
```suggestion
     * K - coefficient, to adjust the loss amount to 0 at the end of the lock period.
```

**Review by ZumZoom — APPROVED** [review]

**Pull request merged**